### DISCUSSION TITLE
Questions on displays and tabs after a compile

### DISCUSSION THREAD

**Discussion by SJLennon** [codefori/vscode-ibmi#1636] [Q&A]
I have not done much on VS Code for a while and the extension has undergone some changes. I wonder if what I'm now seeing when I compile an RPG program is expected. It does not line up with the documentation. Maybe I have some setting that I'm missing.  Or maybe this is by design. 

I'd like to understand, and when I do I'm happy to update the documentation if the changes are stable. (["Compiling"](https://codefori.github.io/docs/#/pages/developing/editing-compiling?id=compiling) is one piece of the documentation that probably need reworking.)

**Here's what I'm seeing:**

I have started  VS Code with no files open. I have opened one tab of an RPG program and run a compile with Ctrl+E action. I see results like this, with a second tab opened up:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/f8a8f82b-669f-4780-8072-e5963cb934b5)

The second tab seems to scroll through the Ctrl+E action as it progresses. The start of the tab looks like this:
```
 *  Executing task: IBM i: Create Bound RPG Program (CRTBNDRPG) 

Current library: LENNONSB
Library list: QIWS QTEMP QGPL LENNONSB LENNONS1
Commands:
                CRTBNDRPG PGM(LENNONS1/AR2) SRCFILE(LENNONS1/TEMP) OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTRLS(*CURRENT)

RNS9308:  Compilation stopped. Severity 30 errors found in program.
CPC0904:  Data area RETURNCODE created in library QTEMP.
RNS9310:  Compilation failed. Program AR2 not created in library LENNONS1.
  5770WDS V7R5M0  220415 RN        IBM ILE RPG             LENNONS1/AR2             PUB400     11/05/23 19:49:01        Page      1
   Command  . . . . . . . . . . . . :   CRTBNDRPG
     Issued by  . . . . . . . . . . :     LENNONS
   Program  . . . . . . . . . . . . :   AR2
     Library  . . . . . . . . . . . :     LENNONS1
   Text 'description' . . . . . . . :   *SRCMBRTXT
   Source Member  . . . . . . . . . :   AR2
   Source File  . . . . . . . . . . :   TEMP
     Library  . . . . . . . . . . . :     LENNONS1
     CCSID  . . . . . . . . . . . . :     273
```

If I click on the "Open Output" button in the notification at the bottom right, a third tab is opened which seems to be a better spaced version of the second tab, except that the first line is removed. (Seems like the [compile listing](https://codefori.github.io/docs/#/pages/settings/global?id=log-compile-output) is no longer found under the "Output" tab and related options are no longer functional.)
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/62858565-8c7b-40e1-b02b-83287eec2048)

If you dismiss the "Open Output" button there is no way to get back other than re-compiling.

I see some discussions about wanting to always open "Problems" automatically after a compile (Post Actions View).  And some desire to see the compile listing, which is occasionally useful to search. 

I'm curious about the rational for the second and third tabs and if they are an attempts to solve this.

<hr />

⚠️ **REMOVE THIS LINE AND ANY SENSITIVE INFORMATION BELOW!** ⚠️

|Context|Version|
|-|-|
|Code for IBM i version|2.4.0|
|Visual Studio Code version|1.83.1|
|Operating System|win32_x64|

<details>
<summary>Active extensions</summary>

```
Code for IBM i Walkthroughs (vscode-ibmi-walkthroughs): 0.4.0
Db2 for IBM i (vscode-db2i): 0.5.1
Emmet (emmet): 1.0.0
Error Lens (errorlens): 3.15.0
Extension Authoring (extension-editing): 1.0.0
Git (git): 1.0.0
Git Base (git-base): 1.0.0
GitHub (github): 0.0.1
GitLens — Git supercharged (gitlens): 14.4.1
IBM i Notebooks (vscode-ibmi-notebooks): 0.0.6
JSON Language Features (json-language-features): 1.0.0
Merge Conflict (merge-conflict): 1.0.0
NPM support for VS Code (npm): 1.0.1
Node Debug Auto-attach (debug-auto-launch): 1.0.0
Overtype (overtype): 0.5.0
RPGLE (vscode-rpgle): 0.23.0
TypeScript and JavaScript Language Features (typescript-language-features): 1.0.0
WSL: Recommender (remote-wsl-recommender): 0.0.19
```
</details>

<hr />

<details>
<summary>Remote system</summary>

|Setting|Value|
|-|-|
|IBM i OS|V7R5M0|
|Tech Refresh|n/a|
|CCSID|37|
|SQL|Enabled
|Source dates|Disabled

### Enabled features

|/QOpenSys/pkgs/bin|/usr/bin|/QSYS.lib/LENNONS2.lib|/QSYS.LIB|/QIBM/ProdData/IBMiDebugService/bin|
|-|-|-|-|-|
|bash|attr|GENCMDXML.PGM|QZDFMDB2.PGM|startDebugService.sh|
|chsh|iconv|GETNEWLIBL.PGM|||
|git|setccsid||||
|grep|||||
|ls|||||
|md5sum|||||
|sort|||||
|stat|||||
|tar|||||
|tn5250|||||
</details>

<details>
<summary>Shell env</summary>

```bash
BUILDLIB=LENNONSB
CURLIB=LENNONSB
HOME=/home/LENNONS
HOST=pub400.com
IFS_LIB=LENNONSB
IFS_SRCPF=VSCODE
LIB1=LENNONSB
LIBLS=LENNONS1 LENNONSB QGPL QTEMP QIWS
LOGIN=lennons
LOGNAME=lennons
MAIL=/var/spool/mail/lennons
OLDPWD=/home/LENNONS
PASE_USRGRP_LIMITED=N
PATH=/QOpenSys/pkgs/bin:/QOpenSys/usr/bin:/usr/ccs/bin:/QOpenSys/usr/bin/X11:/usr/sbin:.:/usr/bin
PWD=/home/LENNONS
SHELL=/QOpenSys/pkgs/bin/bash
SHLVL=1
SSH_CLIENT=71.63.120.199 60705 2222
SSH_CONNECTION=71.63.120.199 60705 185.113.5.134 2222
TZ=<UTC>0
USER=lennons
USERNAME=lennons
_=/QOpenSys/pkgs/bin/env
```
</details>

<details>
<summary>Variants</summary>

```json
{
  "american": "#@$",
  "local": "#§$"
}
```
</details>

<details>
<summary>Errors</summary>

```json
[
  {
    "command": "system \"CRTLIB LIB(LENNONS2) TEXT('Code for i temporary objects. May be cleared.')\"",
    "code": 255,
    "stderr": "CPD0032: Not authorized to command CRTLIB in library *LIBL.\nCPF0006: Errors occurred in command.",
    "cwd": "/home/LENNONS"
  },
  {
    "command": "system \"DLTOBJ OBJ(LENNONS2/O_*) OBJTYPE(*FILE)\"",
    "code": 255,
    "stderr": "CPF2125: No objects deleted.",
    "cwd": "/home/LENNONS"
  }
]
```
</details>

**Comment by worksofliam** [comment]
The 'IBM i Output' channel was removed in favor of a terminal instance to fit in with how VS Code does compiles in other environments. Sometimes the Terminal cuts off everything if the spool result is huge, so the button on the message shows the entire thing no matter the size.

There is a VS Code setting for Code for IBM i that lets users set what they want to see by default after an action has been run.

<img width="615" alt="image" src="https://github.com/codefori/vscode-ibmi/assets/3708366/81f2b8d0-8e6c-4dac-b324-6baa9a44ffdd">

**Comment by SJLennon** [comment]
OK, kind of understand.  
The second editor tab mentioned above is because of this setting, thought how it got set I don't know:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/0f70bb5c-9396-4813-8b6f-174e2e00a0e6)
Changing it to "view" makes more sense, since you don't have an additional editor tab to clutter things up and you can access it only when you need it.

The third editor tab mentioned above, from the "Open output" button. Seems like the button automatically goes away after a few seconds, so there is no need to dismiss it. But it does essentially duplicate the terminal output if you have set "Terminal › Integrated: Scrollback" to a large number. **_So I'm still curious about what problem this solves._** 

I'll see if I can update the documentation.  I suspect many users, like me, rarely use the terminal functionality, and have never investigated  it much, or the 90+ settings it has.

**Comment by SJLennon** [comment]
It appears that whether or not the compile has errors, this message always displays.
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/b72e0fa7-383c-4972-be48-e6c05656d3be)

In past versions if there were errors this was reported with different colored icons (from the help):

![image](https://github.com/codefori/vscode-ibmi/assets/67484051/a721f4c2-e8f4-49ae-b42e-e0c6834df5ae)

If it is no longer possible to distinguish between a successful compile and a failed compile, perhaps the message should say "has completed" instead of "was successful". Though I think it would be preferable to be able to detect and report a failed compile, so folk don't rush off and test with the old version.

**Reply by worksofliam** [reply]
This should not be the case. I still see the error message (with the red X) when a compile fails.

**Reply by SJLennon** [reply]
I have Post Action View set as _problems_ so I then see the errors, as below. 
 ![image](https://github.com/codefori/vscode-ibmi/assets/67484051/b2a4b9f9-8701-4aa4-bb31-e08e41f8f869)
But if it is set to _none_ I get this:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/cf20e1b8-721b-424d-8d9c-94144045c1b1)
Granted, I don't know why you would want anything except problems.

**Reply by SJLennon** [reply]
I have updated the help and will create a PR shortly.  However, since I always get the blue "successful" notification on PUB400 I have not  tried to update the part of the documentation for the red X (this bit: https://codefori.github.io/docs/#/pages/developing/editing-compiling?id=compile-errors). 
Nor have I said anything about the "Open Output" button.

**Comment by SJLennon** [comment]
Don't know if this error I see on all compiles is unique to PUB400 and is related to not seeing the red X:

`qsh: 001-0019 Error found searching for command [[. No such path or directory.`

This is the context:

```
     Comments  . . . . . . . . . . . :        3
          * * * * *   E N D   O F   F I N A L   S U M M A R Y   * * * * *
 Compilation stopped. Severity 30 errors found in program.
           * * * * *   E N D   O F   C O M P I L A T I O N   * * * * *
                                                    -=* http://pub400.com *=-
qsh: 001-0019 Error found searching for command [[. No such path or directory.

Fetching errors for LENNONS1/AR2E.
```

**Reply by worksofliam** [reply]
That would explain why it's always showing successful for you. Is bash your default shell?

**Reply by SJLennon** [reply]
Yes, BASH is my default shell.  No .bash_profile.  

A simple .bashrc. 

![image](https://github.com/codefori/vscode-ibmi/assets/67484051/33ff18f3-bc55-4411-99b8-5afeafcd68a4)

And a .profile:
![image](https://github.com/codefori/vscode-ibmi/assets/67484051/139e78b0-cceb-4fc8-8a07-bbaf498f39be)

**Reply by chrjorgensen** [reply]
@SJLennon The error has been fixed in [PR #1747](https://github.com/codefori/vscode-ibmi/pull/1747) and is included in release 2.6.2.
